tsdate/util.py

```diff
@@ -319,3 +326,268 @@ def add_sampledata_times(samples, sites_time):
+def _split_disjoint_nodes(
+    edges_parent, edges_child, edges_left, edges_right, mutations_edge, nodes_exclude
+):
+    """
+    Split disconnected regions of nodes into separate nodes.
+
+    Returns updated edges_parent, edges_child, mutations_node, and indices indicating
+    from which original node the new nodes are derived.
+    """
+    assert edges_parent.size == edges_child.size == edges_left.size == edges_right.size
+    num_edges = edges_parent.size
+    num_nodes = nodes_exclude.size
+    num_mutations = mutations_edge.size
+
+    # For each edge, check whether parent/child is separated by a gap from the
+    # previous edge involving either parent/child. Label disconnected segments
+    # per node by integers starting at zero.
+    edges_order = np.argsort(edges_left)
+    edges_segments = np.full((2, num_edges), -1, dtype=np.int32)
+    nodes_segments = np.full(num_nodes, -1, dtype=np.int32)
+    nodes_right = np.full(nodes_exclude.size, -np.inf, dtype=np.float64)
+    for e in edges_order:
+        nodes = edges_parent[e], edges_child[e]
+        for i, n in enumerate(nodes):
+            if nodes_exclude[n]:
+                continue
+            nodes_segments[n] += edges_left[e] > nodes_right[n]
+            edges_segments[i, e] = nodes_segments[n]
+            nodes_right[n] = max(nodes_right[n], edges_right[e])
+
+    # Create "nodes_segments[i]" supplementary nodes by copying node "i".
+    # Store the id of the first supplement for each node in "nodes_map".
+    nodes_order = [i for i in range(num_nodes)]
+    nodes_map = np.full(num_nodes, -1, dtype=np.int32)
+    for i, s in enumerate(nodes_segments):
+        for j in range(s):
+            if j == 0:
+                nodes_map[i] = num_nodes
+            nodes_order.append(i)
+            num_nodes += 1
+    nodes_order = np.array(nodes_order, dtype=np.int32)
+
+    # Relabel the nodes on each edge given "nodes_map"
+    for e in edges_order:
+        nodes = edges_parent[e], edges_child[e]
+        for i, n in enumerate(nodes):
+            if edges_segments[i, e] > 0:
+                edges_segments[i, e] += nodes_map[n] - 1
+            else:
+                edges_segments[i, e] = n
+    edges_parent, edges_child = edges_segments[0, ...], edges_segments[1, ...]
+
+    # Relabel node under each mutation
+    mutations_node = np.full(num_mutations, tskit.NULL, dtype=np.int32)
+    for i, e in enumerate(mutations_edge):
+        if e != tskit.NULL:
+            mutations_node[i] = edges_child[e]
+
+    return edges_parent, edges_child, mutations_node, nodes_order
+
+
+def split_disjoint_nodes(ts):
+    """
+    For each non-sample node, split regions separated by gaps into distinct
+    nodes.
+
+    Where there are multiple disconnected regions, the leftmost one is assigned
+    the ID of the original node, and the remainder are assigned new node IDs.
+    Population, flags, individual, time, and metadata are all copied into the
```

Comment:
I think we aren't copying metadata any more, right? just assigning the original node ID into the metadata?